File: practice/combination.js
const arr = ["A", "B", "C", "D"];

const combination = (arr, n) => {
  if (n === 1) return arr.map((v) => [v]);
  const result = [];
  arr.forEach((v, i, origin) => {
    const rest = origin.slice(i + 1);
    console.log("V : ", v, "REST", rest);
    const combinations = combination(rest, n - 1);
    console.log("combinations", combinations);
    const attached = combinations.map((c) => [v, ...c]);
    console.log("attached", attached);
    result.push(...attached);
    console.log("result", result);
  });

  return result;
};

combination(arr, 3);

File: practice/combination.py
# 순서에 상관없이 n개 중에서 r개를 뽑는 경우의 수
# 중복을 허용하지 않는다.
from itertools import combinations
arr = ['A', 'B', 'C', 'D', 'E', 'F']

# 조합 공식
# nCr = n! / r! * (n-r)!
# nCr = n * (n-1) * (n-2) * ... * (n-r+1) / r!
# n의 개수가 6개이고, r의 개수가 3개이므로
# 6C2 = (6 * 5 * 4) / (1 * 2 * 3) = 20

print(len(list(combinations(arr,3)))) 